cpu/esp32/periph/uart.c

```diff
@@ -342,3 +370,74 @@ static int _uart_set_baudrate(uart_t uart, uint32_t baudrate)
     critical_exit();
     return UART_OK;
 }
+
+static int _uart_set_mode(uart_t uart, uart_data_bits_t data_bits,
+                          uart_parity_t parity, uart_stop_bits_t stop_bits)
+{
+    DEBUG("%s uart=%d, data_bits=%d parity=%d stop_bits=%d\n", __func__,
+          uart, data_bits, parity, stop_bits);
+
+    CHECK_PARAM_RET (uart < UART_NUMOF, UART_NODEV);
+
+    critical_enter();
+
+    /* set number of data bits */
+    switch (data_bits) {
+        case UART_DATA_BITS_5: _uarts[uart].regs->conf0.bit_num = 0; break;
+        case UART_DATA_BITS_6: _uarts[uart].regs->conf0.bit_num = 1; break;
+        case UART_DATA_BITS_7: _uarts[uart].regs->conf0.bit_num = 2; break;
+        case UART_DATA_BITS_8: _uarts[uart].regs->conf0.bit_num = 3; break;
+        default: LOG_TAG_ERROR("uart", "invalid number of data bits\n");
+                 critical_exit();
+                 return UART_NOMODE;
+    }
+    /* store changed number of data bits in configuration */
+    _uarts[uart].data = data_bits;
+
+    /* set number of stop bits */
+    #ifdef MCU_ESP32
+    /* workaround for hardware bug when stop bits are set to 2-bit mode. */
+    switch (stop_bits) {
+        case UART_STOP_BITS_1: _uarts[uart].regs->conf0.stop_bit_num = 1;
+                               _uarts[uart].regs->rs485_conf.dl1_en = 1;
+                               break;
+        case UART_STOP_BITS_2: _uarts[uart].regs->conf0.stop_bit_num = 1;
```

Comment:
Should this be 2?  Either that or say not implemented.

Reply:
No, it is a known hardware bug. ` _uarts[uart].regs->conf0.stop_bit_num` has to be set to `1` if there is any stop bit. With `_uarts[uart].regs->rs485_conf.dl1_en` it is decided whether there are 1 or 2 stop bits.